Add magnetic heading as debug and magnetic declination for the Mahony filter

pichim opened the pull request:
After the mag overhaul debug parts from here may be removed.

ledvinap reviewed (commented):
This is overkill. Using matrix multiplication is fine for debug code, but (IMO) not here.

Easiest (and clean) way to get mag declination is to set it in `north_ef`.  It will cost few MACs (north_ef.y ==0 is not optimized).

Another, probably better way would be to use 3x4 matrix (technically 4x4, with [0,0,0,1] as last row. homogenous coordinated math)) to get sensor alignment , gain compensation, offset compensation and now mag declination in single matrix multiplication.

ledvinap reviewed (approved):
I did not test that north_ef direction is correct, but simple live test should do.

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1004,10 +1004,11 @@ const clivalue_t valueTable[] = {
+    { "imu_dcm_ki",                            VAR_UINT16 | MASTER_VALUE, .config.minmaxUnsigned = { 0, 32000 }, PG_IMU_CONFIG, offsetof(imuConfig_t, dcm_ki) },
+    { "small_angle",                           VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 0, 180 }, PG_IMU_CONFIG, offsetof(imuConfig_t, small_angle) },
+    { "imu_process_denom",                     VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 1, 4 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_process_denom) },
+    { "imu_magnetic_declination_deci_degrees", VAR_UINT16 | MASTER_VALUE, .config.minmaxUnsigned = { 0, 3599 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_magnetic_declination_deci_degrees) },
```
Maybe be a good time to add these IMU related parameter names to the `PARAM_NAME` list?
Perhaps `mag_declination_deci_deg` would suffice for a name?

haslinghuis replied:
Or just `mag_declination_dd`

pichim replied:
Ok i tried to add a PARAM_NAME list. But im unsure if everything is done how it should be. Also i don't know if i used the ifdef USE_MAG is used correctly.

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1004,10 +1004,13 @@ const clivalue_t valueTable[] = {
+    { PARAM_NAME_IMU_DCM_KP,          VAR_UINT16 | MASTER_VALUE, .config.minmaxUnsigned = { 0, 32000 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_dcm_kp) },
+    { PARAM_NAME_IMU_DCM_KI,          VAR_UINT16 | MASTER_VALUE, .config.minmaxUnsigned = { 0, 32000 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_dcm_ki) },
+    { PARAM_NAME_IMU_SMALL_ANGLE,     VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 0,   180 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_small_angle) },
+    { PARAM_NAME_PROCESS_DENOM,       VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 1,     4 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_process_denom) },
```
maybe 'PARAM_NAME_IMU_PROCESS_DENOM' to be consistent with the others?

pichim replied:
Thx for pointing that out, i will adjust it.

pichim commented:
Since this PR is based on https://github.com/betaflight/betaflight/pull/13063, should i just close https://github.com/betaflight/betaflight/pull/13063 so that it is not getting too complicated? It's only minor changes anyway i think.

haslinghuis commented on src/main/fc/parameter_names.h:
```diff
@@ -197,3 +197,11 @@
+#define PARAM_NAME_IMU_PROCESS_DENOM "imu_process_denom"
+#ifdef USE_MAG
+#define PARAM_NAME_IMU_MAG_DECLINATION "imu_mag_declination"
+#endif
```
```suggestion
#endif

```

ctzsnooze commented on src/main/fc/parameter_names.h:
```diff
@@ -197,3 +197,11 @@
+#define PARAM_NAME_IMU_SMALL_ANGLE "imu_small_angle"
+#define PARAM_NAME_IMU_PROCESS_DENOM "imu_process_denom"
+#ifdef USE_MAG
+#define PARAM_NAME_IMU_MAG_DECLINATION "imu_mag_declination"
```
Sorry to ask again, however while adjusting mag settings I would go 'get mag' in CLI, and this was the odd one out; it wasn't listed with the other `mag_...` parameters.  Although the code adjusting declination is done within the IMU part of our code base, I wonder if simply calling it `mag_declination` would be more transparent for the average user?

pichim replied:
Sure, good point. Should i just change it to
```
#define PARAM_NAME_IMU_MAG_DECLINATION "mag_declination"
```
so that in CLI it is listed with get mag, but in the codebase it is clear that it belongs to imu?

pichim replied:
maybe i should also change imu_small_angle back to small_angle? a lot of people will see this otherwise when porting.

pichim replied:
Ok, done.

ctzsnooze commented:
It would be great to run the compass task at 200Hz, since that significantly improves performance for the QMC5333L, and will minimise aliasing for Mags with ODR of 10Hz or 15Hz, which would otherwise be sampled asynchronously at 10Hz.  All Mag modules should run as fast as their firmware options permit.

pichim commented:
@ctzsnooze i agree. but i think this is out of scope of this PR.

ctzsnooze commented:
@pichim @ledvinap please see [this comment](https://github.com/betaflight/betaflight/pull/13072#issuecomment-1722584591) in PR #13072 where I describe an issue where `magYaw` and the IMU heading are adversely impacted during pure pitch inputs at constant yaw and roll angle.  The errors appear to develop as a result of the calculations between new Mag data points.

ctzsnooze commented:
Note that I've made a small [blackbox log viewer PR](https://github.com/betaflight/blackbox-log-viewer/pull/656) to properly display `magYaw`.

pichim commented:
Just a basic question about github: Why is noone assigned to this PR?

pichim commented:
added changes from @ctzsnooze so he can rebase gps rescue related work on this. not sure if the ``mag.ìsNewMagADCFlag`` detection will be set and therefor can be detected before alignment and offset removal did take place. changes are work from @ledvinap, @ctzsnooze and @SteveCEvans. i just rebased included it here.

ctzsnooze commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -227,7 +228,7 @@ static bool hmc5883lInit(magDev_t *mag)
 
     // leave test mode
-    busWriteRegister(dev, HMC58X3_REG_CONFA, HMC_CONFA_8_SAMLES | HMC_CONFA_DOR_15HZ | HMC_CONFA_NORMAL);    // Configuration Register A  -- 0 11 100 00  num samples: 8 ; output rate: 15Hz ; normal measurement mode
+    busWriteRegister(dev, HMC58X3_REG_CONFA, HMC_CONFA_8_SAMLES | HMC_CONFA_DOR_75HZ | HMC_CONFA_NORMAL);    // Configuration Register A  -- 0 11 100 00  num samples: 8 ; output rate: 15Hz ; normal measurement mode
```
Comment should be:

Configuration Register A -- 0 11 110 00 num samples: 8 ; output rate: 75Hz ; normal measurement mode

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1003,16 +1003,18 @@ const clivalue_t valueTable[] = {
+    { PARAM_NAME_IMU_SMALL_ANGLE,     VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 0,   180 }, PG_IMU_CONFIG, offsetof(imuConfig_t, small_angle) },
+    { PARAM_NAME_IMU_PROCESS_DENOM,   VAR_UINT8  | MASTER_VALUE, .config.minmaxUnsigned = { 1,     4 }, PG_IMU_CONFIG, offsetof(imuConfig_t, imu_process_denom) },
+#ifdef USE_MAG
+    { PARAM_NAME_IMU_MAG_DECLINATION, VAR_UINT16 | MASTER_VALUE, .config.minmaxUnsigned = { 0,  3599 }, PG_IMU_CONFIG, offsetof(imuConfig_t, mag_declination) },
```
If the declination is negative, eg -5.0 degrees, should this be set as `3550`?  Or should we make this value a signed integer and allow negative values?

ctzsnooze commented:
Thanks, @pichim.  This is looking really good now; all the Mag changes are in this single PR.  

PF#13072 is now quite separate, and only concerns the application of Mag data within GPS Rescue.

Summary of changes in this PR:
- compass Task runs at 200Hz and flags new Mag data arrivals by checking for differences in any one of the axes.
- `magYaw` (mag heading) value, with roll and pitch correction, is now calculated in imu.c when new Mag data arrives. It is logged in the `GPS_RESCUE_HEADING` debug, where it can be readily compared to `attitude.values.yaw` and GPS `Course Over Ground` values for debugging purposes.  This maths doesn't run unless the debug is selected.
- `mag_declination` value in CLI, to set local magnetic declination value, correcting the Mag to read true north, simplifying comparison to, and integration with, course over ground values.
- ODR of the HMC5883L increased to 75Hz
- minor code refactoring, some parameter names put into `parameter_names.h` etc.

I've tested this PR since the most recent update, and it performs extremely well, whether the Mag runs at 10Hz or 200Hz.  A mashup with this PR, #13072 and #13091 provides amazingly solid GPS rescue performance.

Bench testing this PR, while armed, we see, in the top panel, `magYaw` in purple, and `attitude_values_yaw` in yellow.  

The following show tests with a 10Hz Mag unit.  Debug is `GPS_RESCUE_HEADING` and Mag is enabled with `gps_rescue_use_mag = ON`.  Note very little delay between the IMU heading estimate and the new mag data:

![Screen Shot 2023-09-25 at 18 55 39](https://github.com/betaflight/betaflight/assets/11737748/5a474798-55ee-4cea-aef7-50f229ae18ac)

![Screen Shot 2023-09-25 at 18 54 49](https://github.com/betaflight/betaflight/assets/11737748/ee47693d-ca65-44ed-8e8d-bd86662121d0)

These images show similar tests with a QMC5883L with its default ODR of 200hz.  Obviously much more granular and detailed heading information.  The small static offsets arose from the motors being in slightly different positions as I moved it around, I don't see them when there are no motors on the frame.  The second image shows quick pure yaw inputs.

![Screen Shot 2023-09-25 at 15 22 50](https://github.com/betaflight/betaflight/assets/11737748/0d2660eb-6bc2-458c-ba0a-5510ecd05e96)

![Screen Shot 2023-09-25 at 14 44 03](https://github.com/betaflight/betaflight/assets/11737748/d0714f66-3005-4f18-934d-ae7565a0526c)

Overall, performance is truly excellent now.  Definitely ready for final review.

For information on calibration and location of Mag units,[ please see the wiki note](https://betaflight.com/docs/wiki/archive/Magnetometer).

ledvinap commented:
From the traces, time constant of mag alignment is much larger than movement in images (offset between mag and heading is almost constant). High frequency mag is great for visuals, but has negligible influence on actual IMU performance. It will only manifest after multiple fast spins in one direction
